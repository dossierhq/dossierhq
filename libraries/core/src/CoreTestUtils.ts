import { Temporal } from '@js-temporal/polyfill';
import type { ErrorType, Logger, OkResult, Result } from '.';

export function expectOkResult<TOk, TError extends ErrorType>(
  actual: Result<unknown, ErrorType>
): actual is OkResult<TOk, TError> {
  if (actual.isError()) {
    throw new Error(`Expected ok, got error ${actual.error}: ${actual.message}`);
  }
  return true;
}

export function expectErrorResult(
  actual: Result<unknown, ErrorType>,
  expectedErrorType: ErrorType,
  expectedMessage: string
): void {
  if (!actual.isError()) {
    throw new Error(`Expected error, but was ok`);
  }
  const expectedString = `${expectedErrorType}: ${expectedMessage}`;
  const actualString = `${actual.error}: ${actual.message}`;
  if (actualString !== expectedString) {
    throw new Error(`Expected (${expectedString}), got ${actualString}`);
  }
}

export function expectResultValue<TOk, TError extends ErrorType>(
  result: Result<TOk, TError>,
  expectedValue: TOk
): void {
  if (expectOkResult(result)) {
    const actualCopy = deepCopyForIsEqual(result.value);
    const expectedCopy = deepCopyForIsEqual(expectedValue);
    expect(actualCopy).toEqual<TOk>(expectedCopy);
  }
}

function deepCopyForIsEqual<T>(obj: T): T {
  if (obj === null || obj === undefined) return obj;
  if (obj instanceof Temporal.Instant) {
    // Since the epoch isn't stored as a property for Instant (but a slot), jest isn't able to compare them.
    // Replace with string representation
    return obj.toString() as unknown as T;
  }
  if (typeof obj === 'object') {
    const copy = { ...obj };
    for (const [key, value] of Object.entries(obj)) {
      copy[key as keyof T] = deepCopyForIsEqual(value);
    }
    return copy;
  }
  return obj;
}

export function createMockLogger(): {
  error: jest.MockedFunction<Logger['error']>;
  warn: jest.MockedFunction<Logger['warn']>;
  info: jest.MockedFunction<Logger['info']>;
  debug: jest.MockedFunction<Logger['debug']>;
} {
  return {
    error: jest.fn(),
    warn: jest.fn(),
    info: jest.fn(),
    debug: jest.fn(),
  };
}
