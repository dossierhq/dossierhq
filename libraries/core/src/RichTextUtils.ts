import { RichTextNodeType } from './Schema.js';
import type {
  HeadingTagType,
  SerializedHeadingNode,
  SerializedParagraphNode,
  SerializedTextNode,
  TextFormatType,
} from './third-party/Lexical.js';
import { TEXT_TYPE_TO_FORMAT } from './third-party/Lexical.js';
import type {
  EntityReference,
  RichText,
  RichTextEntityLinkNode,
  RichTextEntityNode,
  RichTextNode,
  RichTextValueItemNode,
  ValueItem,
} from './Types.js';

export function createRichTextRootNode(children: RichTextNode[]): RichText {
  return {
    root: {
      direction: 'ltr',
      format: '',
      indent: 0,
      type: 'root',
      version: 1,
      children,
    },
  };
}

export function createRichTextHeadingNode(
  tag: HeadingTagType,
  children: RichTextNode[]
): SerializedHeadingNode {
  return {
    direction: 'ltr',
    format: '',
    indent: 0,
    tag,
    type: 'heading',
    version: 1,
    children,
  };
}

export function createRichTextParagraphNode(children: RichTextNode[]): SerializedParagraphNode {
  return {
    direction: 'ltr',
    format: '',
    indent: 0,
    type: 'paragraph',
    version: 1,
    children,
  };
}

export function createRichTextTextNode(
  text: string,
  options?: { format?: TextFormatType[] }
): SerializedTextNode {
  let formatValue = 0;
  if (options?.format) {
    for (const formatItem of options.format) {
      const formatFlag = TEXT_TYPE_TO_FORMAT[formatItem];
      if (formatFlag !== undefined) {
        formatValue |= formatFlag;
      }
    }
  }
  return {
    detail: 0,
    format: formatValue,
    mode: 'normal',
    style: '',
    text,
    type: 'text',
    version: 1,
  };
}

export function createRichTextEntityNode(reference: EntityReference): RichTextEntityNode {
  return { type: RichTextNodeType.entity, reference, version: 1 };
}

export function createRichTextEntityLinkNode(
  reference: EntityReference,
  children: RichTextNode[]
): RichTextEntityLinkNode {
  return {
    direction: 'ltr',
    format: '',
    indent: 0,
    type: RichTextNodeType.entityLink,
    reference,
    version: 1,
    children,
  };
}

export function createRichTextValueItemNode<T extends ValueItem<string, object>>(
  data: T
): RichTextValueItemNode {
  return { type: RichTextNodeType.valueItem, data, version: 1 };
}
