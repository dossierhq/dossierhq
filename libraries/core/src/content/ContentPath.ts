export type ContentValuePath = (string | number)[];

export function visitorPathToString(path: ContentValuePath): string {
  let result = '';
  for (const segment of path) {
    if (Number.isInteger(segment)) {
      result += `[${segment}]`;
    } else {
      if (result.length === 0) {
        result += segment;
      } else {
        result += `.${segment}`;
      }
    }
  }
  return result;
}
