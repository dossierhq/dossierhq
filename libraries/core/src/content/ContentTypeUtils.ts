import type {
  AdminEntity,
  AdminEntityCreate,
  AdminEntityUpdate,
  Component,
  EntityLike,
  PublishedEntity,
  RichTextCodeHighlightNode,
  RichTextCodeNode,
  RichTextElementNode,
  RichTextEntityLinkNode,
  RichTextEntityNode,
  RichTextHeadingNode,
  RichTextLineBreakNode,
  RichTextLinkNode,
  RichTextListItemNode,
  RichTextListNode,
  RichTextNode,
  RichTextParagraphNode,
  RichTextRootNode,
  RichTextTabNode,
  RichTextTextNode,
  RichTextComponentNode,
} from '../Types.js';
import { RichTextNodeType } from '../Types.js';
import type { FieldSpecification, FieldValueTypeMap } from '../schema/SchemaSpecification.js';
import { FieldType } from '../schema/SchemaSpecification.js';

type WithRichTextType<TNode extends RichTextNode, TType extends RichTextNodeType> = Omit<
  TNode,
  'type'
> & { type: TType };

/** Check if `value` with `fieldSpec` is a single boolean field */
export function isBooleanSingleField(
  fieldSpec: FieldSpecification,
  value: unknown,
): value is FieldValueTypeMap[typeof FieldType.Boolean] | null {
  return fieldSpec.type === FieldType.Boolean && !fieldSpec.list;
}

/** Check if `value` with `fieldSpec` is a list boolean field */
export function isBooleanListField(
  fieldSpec: FieldSpecification,
  value: unknown,
): value is FieldValueTypeMap[typeof FieldType.Boolean][] | null {
  return fieldSpec.type === FieldType.Boolean && fieldSpec.list;
}

/** Check if `value` with `fieldSpec` is either a single boolean field or an item in a list field */
export function isBooleanItemField(
  fieldSpec: FieldSpecification,
  value: unknown,
): value is FieldValueTypeMap[typeof FieldType.Boolean] | null {
  return fieldSpec.type === FieldType.Boolean;
}

/** Check if `value` with `fieldSpec` is a single Component field */
export function isComponentSingleField(
  fieldSpec: FieldSpecification,
  value: unknown,
): value is FieldValueTypeMap[typeof FieldType.Component] | null {
  return fieldSpec.type === FieldType.Component && !fieldSpec.list;
}

/** Check if `value` with `fieldSpec` is a list Component field */
export function isComponentListField(
  fieldSpec: FieldSpecification,
  value: unknown,
): value is FieldValueTypeMap[typeof FieldType.Component][] | null {
  return fieldSpec.type === FieldType.Component && fieldSpec.list;
}

/** Check if `value` with `fieldSpec` is either a single Component field or an item in a list field */
export function isComponentItemField(
  fieldSpec: FieldSpecification,
  value: unknown,
): value is FieldValueTypeMap[typeof FieldType.Component] | null {
  return fieldSpec.type === FieldType.Component;
}

/** Check if `value` with `fieldSpec` is a single Entity field */
export function isEntitySingleField(
  fieldSpec: FieldSpecification,
  value: unknown,
): value is FieldValueTypeMap[typeof FieldType.Entity] | null {
  return fieldSpec.type === FieldType.Entity && !fieldSpec.list;
}

/** Check if `value` with `fieldSpec` is a list Entity field */
export function isEntityListField(
  fieldSpec: FieldSpecification,
  value: unknown,
): value is FieldValueTypeMap[typeof FieldType.Entity][] | null {
  return fieldSpec.type === FieldType.Entity && fieldSpec.list;
}

/** Check if `value` with `fieldSpec` is either a single Entity field or an item in a list field */
export function isEntityItemField(
  fieldSpec: FieldSpecification,
  value: unknown,
): value is FieldValueTypeMap[typeof FieldType.Entity] | null {
  return fieldSpec.type === FieldType.Entity;
}

/** Check if `value` with `fieldSpec` is a single Location field */
export function isLocationSingleField(
  fieldSpec: FieldSpecification,
  value: unknown,
): value is FieldValueTypeMap[typeof FieldType.Location] | null {
  return fieldSpec.type === FieldType.Location && !fieldSpec.list;
}

/** Check if `value` with `fieldSpec` is a list Location field */
export function isLocationListField(
  fieldSpec: FieldSpecification,
  value: unknown,
): value is FieldValueTypeMap[typeof FieldType.Location][] | null {
  return fieldSpec.type === FieldType.Location && fieldSpec.list;
}

/** Check if `value` with `fieldSpec` is either a single Location field or an item in a list field */
export function isLocationItemField(
  fieldSpec: FieldSpecification,
  value: unknown,
): value is FieldValueTypeMap[typeof FieldType.Location] | null {
  return fieldSpec.type === FieldType.Location;
}

/** Check if `value` with `fieldSpec` is a single number field */
export function isNumberSingleField(
  fieldSpec: FieldSpecification,
  value: unknown,
): value is FieldValueTypeMap[typeof FieldType.Number] | null {
  return fieldSpec.type === FieldType.Number && !fieldSpec.list;
}

/** Check if `value` with `fieldSpec` is a list number field */
export function isNumberListField(
  fieldSpec: FieldSpecification,
  value: unknown,
): value is FieldValueTypeMap[typeof FieldType.Number][] | null {
  return fieldSpec.type === FieldType.Number && fieldSpec.list;
}

/** Check if `value` with `fieldSpec` is either a single number field or an item in a list field */
export function isNumberItemField(
  fieldSpec: FieldSpecification,
  value: unknown,
): value is FieldValueTypeMap[typeof FieldType.Number] | null {
  return fieldSpec.type === FieldType.Number;
}

/** Check if `value` with `fieldSpec` is a single String field */
export function isStringSingleField(
  fieldSpec: FieldSpecification,
  value: unknown,
): value is FieldValueTypeMap[typeof FieldType.String] | null {
  return fieldSpec.type === FieldType.String && !fieldSpec.list;
}

/** Check if `value` with `fieldSpec` is a list String field */
export function isStringListField(
  fieldSpec: FieldSpecification,
  value: unknown,
): value is FieldValueTypeMap[typeof FieldType.String][] | null {
  return fieldSpec.type === FieldType.String && fieldSpec.list;
}

/** Check if `value` with `fieldSpec` is either a single String field or an item in a list field */
export function isStringItemField(
  fieldSpec: FieldSpecification,
  value: unknown,
): value is FieldValueTypeMap[typeof FieldType.String] | null {
  return fieldSpec.type === FieldType.String;
}

/** Check if `value` with `fieldSpec` is a single RichText field */
export function isRichTextSingleField(
  fieldSpec: FieldSpecification,
  value: unknown,
): value is FieldValueTypeMap[typeof FieldType.RichText] | null {
  return fieldSpec.type === FieldType.RichText && !fieldSpec.list;
}

/** Check if `value` with `fieldSpec` is a list RichText field */
export function isRichTextListField(
  fieldSpec: FieldSpecification,
  value: unknown,
): value is FieldValueTypeMap[typeof FieldType.RichText][] | null {
  return fieldSpec.type === FieldType.RichText && fieldSpec.list;
}

/** Check if `value` with `fieldSpec` is either a single RichText field or an item in a list field */
export function isRichTextItemField(
  fieldSpec: FieldSpecification,
  value: unknown,
): value is FieldValueTypeMap[typeof FieldType.RichText] | null {
  return fieldSpec.type === FieldType.RichText;
}

export function isRichTextTextNode(
  node: RichTextNode,
): node is WithRichTextType<RichTextTextNode, 'text'> {
  return node.type === RichTextNodeType.text;
}

export function isRichTextLineBreakNode(
  node: RichTextNode,
): node is WithRichTextType<RichTextLineBreakNode, 'linebreak'> {
  return node.type === RichTextNodeType.linebreak;
}

export function isRichTextTabNode(
  node: RichTextNode,
): node is WithRichTextType<RichTextTabNode, 'tab'> {
  return node.type === RichTextNodeType.tab;
}

export function isRichTextElementNode(node: RichTextNode): node is RichTextElementNode {
  return 'children' in node;
}

export function isRichTextRootNode(node: RichTextNode): node is RichTextRootNode {
  return node.type === RichTextNodeType.root;
}

export function isRichTextParagraphNode(
  node: RichTextNode,
): node is WithRichTextType<RichTextParagraphNode, 'paragraph'> {
  return node.type === RichTextNodeType.paragraph;
}

export function isRichTextComponentNode(
  node: RichTextNode,
): node is WithRichTextType<RichTextComponentNode, 'component'> {
  return node.type === RichTextNodeType.component;
}

export function isRichTextEntityNode(
  node: RichTextNode,
): node is WithRichTextType<RichTextEntityNode, 'entity'> {
  return node.type === RichTextNodeType.entity;
}

export function isRichTextEntityLinkNode(
  node: RichTextNode,
): node is WithRichTextType<RichTextEntityLinkNode, 'entityLink'> {
  return node.type === RichTextNodeType.entityLink;
}

export function isRichTextCodeNode(
  node: RichTextNode,
): node is WithRichTextType<RichTextCodeNode, 'code'> {
  return node.type === RichTextNodeType.code;
}

export function isRichTextCodeHighlightNode(
  node: RichTextNode,
): node is WithRichTextType<RichTextCodeHighlightNode, 'code-highlight'> {
  return node.type === RichTextNodeType['code-highlight'];
}

export function isRichTextHeadingNode(
  node: RichTextNode,
): node is WithRichTextType<RichTextHeadingNode, 'heading'> {
  return node.type === RichTextNodeType.heading;
}

export function isRichTextLinkNode(
  node: RichTextNode,
): node is WithRichTextType<RichTextLinkNode, 'link'> {
  return node.type === RichTextNodeType.link;
}

export function isRichTextListNode(
  node: RichTextNode,
): node is WithRichTextType<RichTextListNode, 'list'> {
  return node.type === RichTextNodeType.list;
}

export function isRichTextListItemNode(
  node: RichTextNode,
): node is WithRichTextType<RichTextListItemNode, 'listitem'> {
  return node.type === RichTextNodeType.listitem;
}

export function isComponent(
  item:
    | Component
    | PublishedEntity
    | AdminEntity
    | AdminEntityCreate
    | AdminEntityUpdate
    | EntityLike,
): item is Component {
  return 'type' in item;
}

export function isAdminEntity(
  item: Component | PublishedEntity | AdminEntity,
): item is AdminEntity {
  return !isComponent(item) && 'version' in item.info;
}

export function isPublishedEntity(
  item: Component | PublishedEntity | AdminEntity,
): item is PublishedEntity {
  return !isComponent(item) && !isAdminEntity(item);
}
