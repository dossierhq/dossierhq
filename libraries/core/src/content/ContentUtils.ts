import type {
  AdminEntity,
  AdminEntityCreate,
  EntityLike,
  Location,
  PublishedEntity,
  RichText,
  RichTextNode,
} from '../Types.js';
import { FieldType, type FieldSpecification } from '../schema/SchemaSpecification.js';
import { assertExhaustive } from '../utils/Asserts.js';
import type { Mutable } from '../utils/TypeUtils.js';
import type { ContentValuePath } from './ContentPath.js';
import { isRichTextRootNode } from './ContentTypeUtils.js';

export function getEntityNameBase(name: string): string {
  const hashIndex = name.lastIndexOf('#');
  if (hashIndex < 0) {
    return name;
  }
  return name.slice(0, hashIndex);
}

export function isEntityNameAsRequested(currentName: string, requestedName: string): boolean {
  if (requestedName === currentName) {
    return true;
  }
  const currentNameBase = getEntityNameBase(currentName);
  return requestedName === currentNameBase;
}

export function copyEntity<
  T extends
    | AdminEntity<string, object>
    | AdminEntityCreate<AdminEntity<string, object>>
    | PublishedEntity<string, object>
    | EntityLike<string, object>,
>(
  entity: Readonly<T> | T,
  changes: { id?: string; info?: Partial<T['info']>; fields?: Partial<T['fields']> },
): Readonly<T> {
  const copy: Mutable<T> = { ...entity };
  if (typeof changes.id === 'string') {
    copy.id = changes.id;
  }
  if (changes.info) {
    copy.info = { ...entity.info };
    for (const [key, value] of Object.entries(changes.info)) {
      (copy.info as unknown as Record<string, unknown>)[key] = value;
    }
  }
  if (changes.fields) {
    const fieldsCopy = { ...entity.fields } as Record<string, unknown>;
    copy.fields = fieldsCopy;
    for (const [key, value] of Object.entries(changes.fields)) {
      fieldsCopy[key] = value;
    }
  }
  return copy;
}

export function checkFieldTraversable(
  fieldSpec: FieldSpecification,
  value: unknown,
): { path: ContentValuePath; message: string } | null {
  if (fieldSpec.list) {
    if (value !== null && value !== undefined && !Array.isArray(value)) {
      return { path: [], message: `Expected a list of ${fieldSpec.type}, got ${typeof value}` };
    }
  }
  return null;
}

export function checkFieldItemTraversable(
  fieldSpec: FieldSpecification,
  value: unknown,
): { path: ContentValuePath; message: string } | null {
  const { type } = fieldSpec;

  if (value === null || value === undefined) {
    return null;
  }

  if (Array.isArray(value)) {
    return { path: [], message: `Expected single ${type}, got a list` };
  }

  switch (type) {
    case FieldType.Boolean: {
      if (typeof value !== 'boolean') {
        return { path: [], message: `Expected a boolean, got ${typeof value}` };
      }
      break;
    }
    case FieldType.Component: {
      if (typeof value !== 'object') {
        return { path: [], message: `Expected a Component object, got ${typeof value}` };
      }
      if (!('type' in value)) {
        return { path: ['type'], message: `Missing a Component type` };
      }
      if (typeof value.type !== 'string') {
        return { path: ['type'], message: `Expected a Component type, got ${typeof value}` };
      }
      break;
    }
    case FieldType.Reference: {
      if (typeof value !== 'object') {
        return { path: [], message: `Expected an entity reference, got ${typeof value}` };
      }
      if (typeof (value as EntityLike).id !== 'string') {
        return {
          path: ['id'],
          message: `Expected an entity reference id, got ${typeof (value as EntityLike).id}`,
        };
      }
      break;
    }
    case FieldType.Location: {
      if (typeof value !== 'object') {
        return { path: [], message: `Expected a Location object, got ${typeof value}` };
      }
      const { lat, lng } = value as Location;
      if (typeof lat !== 'number' || typeof lng !== 'number') {
        return {
          path: [],
          message: `Expected {lat: number, lng: number}, got {lat: ${typeof lat}, lng: ${typeof lng}}`,
        };
      }
      break;
    }
    case FieldType.Number: {
      if (typeof value !== 'number') {
        return { path: [], message: `Expected a number, got ${typeof value}` };
      }
      break;
    }
    case FieldType.RichText: {
      if (typeof value !== 'object') {
        return { path: [], message: `Expected a RichText object, got ${typeof value}` };
      }
      const root = (value as RichText).root;

      if (!root) {
        return { path: [], message: `RichText object is missing root` };
      }

      const rootTraversalError = checkRichTextNodeTraversable(root);
      if (rootTraversalError) {
        return { path: ['root', ...rootTraversalError.path], message: rootTraversalError.message };
      }

      if (!isRichTextRootNode(root)) {
        return {
          path: ['root'],
          message: `RichText root is not a valid RichText node, (got ${
            (root as RichTextNode).type
          })`,
        };
      }
      break;
    }
    case FieldType.String: {
      if (typeof value !== 'string') {
        return { path: [], message: `Expected a string, got ${typeof value}` };
      }
      break;
    }
    default:
      assertExhaustive(type);
  }
  return null;
}

export function checkRichTextNodeTraversable(
  node: RichTextNode,
): { path: ContentValuePath; message: string } | null {
  if (!node || typeof node !== 'object') {
    return { path: [], message: `Expected a RichText node, got ${typeof node}` };
  }
  if (typeof node.type !== 'string') {
    return { path: [], message: `RichText node is missing type` };
  }
  if ('children' in node && !Array.isArray(node.children)) {
    return { path: [], message: `RichText node children is not an array` };
  }
  return null;
}
