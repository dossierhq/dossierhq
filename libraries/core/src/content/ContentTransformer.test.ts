import { describe, expect, test } from 'vitest';
import { ErrorType, ok } from '../ErrorResult.js';
import type { Component } from '../Types.js';
import { createRichText, createRichTextValueItemNode } from '../content/RichTextUtils.js';
import { AdminSchemaWithMigrations } from '../schema/AdminSchema.js';
import { FieldType } from '../schema/SchemaSpecification.js';
import { expectErrorResult } from '../test/CoreTestUtils.js';
import { contentValuePathToString } from './ContentPath.js';
import {
  IDENTITY_TRANSFORMER,
  transformEntityFields,
  transformComponent,
} from './ContentTransformer.js';
import { isRichTextValueItemNode, isComponentItemField } from './ContentTypeUtils.js';
import { copyEntity } from './ContentUtils.js';

const ADMIN_SCHEMA = AdminSchemaWithMigrations.createAndValidate({
  entityTypes: [
    {
      name: 'StringsEntity',
      fields: [
        { name: 'string', type: FieldType.String },
        { name: 'stringList', type: FieldType.String, list: true },
      ],
    },
    {
      name: 'ComponentsEntity',
      fields: [
        { name: 'richText', type: FieldType.RichText },
        { name: 'component', type: FieldType.Component },
        { name: 'componentList', type: FieldType.Component, list: true },
      ],
    },
  ],
  componentTypes: [
    {
      name: 'NestedComponent',
      fields: [
        { name: 'child', type: FieldType.Component },
        { name: 'string', type: FieldType.String },
      ],
    },
  ],
}).valueOrThrow();

const STRINGS_ENTITY_1 = Object.freeze({
  info: { type: 'StringsEntity' },
  fields: {
    string: '1',
    stringList: ['1', '2'],
  },
});

const COMPONENTS_ENTITY_1 = Object.freeze({
  info: { type: 'ComponentsEntity' },
  fields: {
    richText: createRichText([
      createRichTextValueItemNode({ type: 'NestedComponent', child: null, string: null }),
    ]),
    component: {
      type: 'NestedComponent',
      string: '1',
      child: { type: 'NestedComponent', string: '1.1', child: null },
    },
    componentList: [
      { type: 'NestedComponent', string: '1', child: null },
      { type: 'NestedComponent', string: '2', child: null },
    ],
  },
});

describe('transformEntity', () => {
  test('identity', () => {
    const calls: unknown[][] = [];
    const transformed = transformEntityFields(ADMIN_SCHEMA, [], COMPONENTS_ENTITY_1, {
      transformField: (_schema, path, _fieldSpec, value) => {
        calls.push(['transformField', contentValuePathToString(path)]);
        return ok(value);
      },
      transformFieldItem: (_schema, path, _fieldSpec, value) => {
        calls.push(['transformFieldItem', contentValuePathToString(path)]);
        return ok(value);
      },
      transformRichTextNode: (_schema, path, _fieldSpec, node) => {
        calls.push(['transformRichTextNode', contentValuePathToString(path)]);
        return ok(node);
      },
    }).valueOrThrow();
    expect(transformed).toBe(COMPONENTS_ENTITY_1.fields);
    expect(calls).toMatchSnapshot();
  });

  test('delete all components', () => {
    const transformed = transformEntityFields(ADMIN_SCHEMA, [], COMPONENTS_ENTITY_1, {
      transformField: (_schema, _path, _fieldSpec, value) => ok(value),
      transformFieldItem: (_schema, _path, fieldSpec, value) => {
        if (isComponentItemField(fieldSpec, value)) return ok(null);
        return ok(value);
      },
      transformRichTextNode: (_schema, _path, _fieldSpec, node) =>
        ok(isRichTextValueItemNode(node) ? null : node),
    }).valueOrThrow();
    expect(transformed).toMatchSnapshot();
  });

  test('normalize entity fields: extra', () => {
    const transformed = transformEntityFields(
      ADMIN_SCHEMA,
      ['entity'],
      copyEntity(STRINGS_ENTITY_1, {
        fields: { extra: 'hello' } as unknown as typeof STRINGS_ENTITY_1.fields,
      }),
      IDENTITY_TRANSFORMER,
    ).valueOrThrow();
    expect(transformed).toEqual({ string: '1', stringList: ['1', '2'] });
  });

  test('normalize entity fields: extra with keepExtraFields', () => {
    const transformed = transformEntityFields(
      ADMIN_SCHEMA,
      ['entity'],
      copyEntity(STRINGS_ENTITY_1, {
        fields: { extra: 'hello' } as unknown as typeof STRINGS_ENTITY_1.fields,
      }),
      IDENTITY_TRANSFORMER,
      { keepExtraFields: true },
    ).valueOrThrow();
    expect(transformed).toEqual({ extra: 'hello', string: '1', stringList: ['1', '2'] });
  });

  test('normalize list: empty', () => {
    const transformed = transformEntityFields(
      ADMIN_SCHEMA,
      ['entity'],
      copyEntity(STRINGS_ENTITY_1, { fields: { stringList: [] } }),
      IDENTITY_TRANSFORMER,
    ).valueOrThrow();
    expect(transformed.stringList).toBe(null);
  });

  test('normalize string: empty', () => {
    const transformed = transformEntityFields(
      ADMIN_SCHEMA,
      ['entity'],
      copyEntity(STRINGS_ENTITY_1, { fields: { string: '' } }),
      IDENTITY_TRANSFORMER,
    ).valueOrThrow();
    expect(transformed.string).toBe(null);
  });

  test('normalize component: extra fields', () => {
    const copy = copyEntity(COMPONENTS_ENTITY_1, {
      fields: {
        component: {
          type: 'NestedComponent',
          unsupported: 'hello',
        } as unknown as typeof COMPONENTS_ENTITY_1.fields.component,
      },
    });
    const transformed = transformEntityFields(
      ADMIN_SCHEMA,
      ['entity'],
      copy,
      IDENTITY_TRANSFORMER,
    ).valueOrThrow();
    expect(transformed.component).toEqual({ type: 'NestedComponent', child: null, string: null });
  });

  test('normalize component: extra fields with keepExtraFields', () => {
    const copy = copyEntity(COMPONENTS_ENTITY_1, {
      fields: {
        component: {
          type: 'NestedComponent',
          unsupported: 'hello',
        } as unknown as typeof COMPONENTS_ENTITY_1.fields.component,
      },
    });
    const transformed = transformEntityFields(
      ADMIN_SCHEMA,
      ['entity'],
      copy,
      IDENTITY_TRANSFORMER,
      { keepExtraFields: true },
    ).valueOrThrow();
    expect(transformed.component).toEqual({
      type: 'NestedComponent',
      child: null,
      string: null,
      unsupported: 'hello',
    });
  });

  test('error: invalid entity type name', () => {
    const result = transformEntityFields(
      ADMIN_SCHEMA,
      ['entity'],
      copyEntity(STRINGS_ENTITY_1, { info: { type: 'Invalid' } }),
      IDENTITY_TRANSFORMER,
    );
    expectErrorResult(
      result,
      ErrorType.BadRequest,
      'entity: Couldn’t find spec for entity type Invalid',
    );
  });

  test('error: missing type in component', () => {
    const copy = copyEntity(COMPONENTS_ENTITY_1, {
      fields: {
        component: {} as unknown as typeof COMPONENTS_ENTITY_1.fields.component,
      },
    });
    const result = transformEntityFields(ADMIN_SCHEMA, ['entity'], copy, IDENTITY_TRANSFORMER);
    expectErrorResult(
      result,
      ErrorType.BadRequest,
      'entity.component.type: Missing a Component type',
    );
  });

  test('error: string[] where string is expected', () => {
    const copy = copyEntity(STRINGS_ENTITY_1, {
      fields: {
        string: ['one', 'two'] as unknown as typeof STRINGS_ENTITY_1.fields.string,
      },
    });
    const result = transformEntityFields(ADMIN_SCHEMA, ['entity'], copy, IDENTITY_TRANSFORMER);
    expectErrorResult(
      result,
      ErrorType.BadRequest,
      'entity.string: Expected single String, got a list',
    );
  });
});

describe('transformComponent', () => {
  test('normalize component: extra field', () => {
    const transformed = transformComponent(
      ADMIN_SCHEMA,
      ['component'],
      { type: 'NestedComponent', child: null, string: null, extra: 'hello' },
      IDENTITY_TRANSFORMER,
    ).valueOrThrow();
    expect(transformed).toEqual({ type: 'NestedComponent', child: null, string: null });
  });

  test('normalize component: extra field with keepExtraFields', () => {
    const transformed = transformComponent(
      ADMIN_SCHEMA,
      ['component'],
      { type: 'NestedComponent', child: null, string: null, extra: 'hello' },
      IDENTITY_TRANSFORMER,
      { keepExtraFields: true },
    ).valueOrThrow();
    expect(transformed).toEqual({
      type: 'NestedComponent',
      child: null,
      string: null,
      extra: 'hello',
    });
  });

  test('error: invalid type name', () => {
    const transformed = transformComponent(
      ADMIN_SCHEMA,
      ['component'],
      { type: 'Invalid' },
      IDENTITY_TRANSFORMER,
    );
    expectErrorResult(
      transformed,
      ErrorType.BadRequest,
      'component: Couldn’t find spec for component type Invalid',
    );
  });

  test('error: missing type', () => {
    const transformed = transformComponent(
      ADMIN_SCHEMA,
      ['component'],
      {} as Component,
      IDENTITY_TRANSFORMER,
    );
    expectErrorResult(
      transformed,
      ErrorType.BadRequest,
      'component.type: Missing a component type',
    );
  });
});
