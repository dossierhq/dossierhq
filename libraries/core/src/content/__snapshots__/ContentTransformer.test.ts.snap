// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`transformEntity > delete all value items 1`] = `
{
  "richText": null,
  "valueItem": null,
  "valueItemList": null,
}
`;

exports[`transformEntity > identity 1`] = `
[
  [
    "transformField",
    "richText",
  ],
  [
    "transformFieldItem",
    "richText",
  ],
  [
    "transformRichTextNode",
    "richText",
  ],
  [
    "transformRichTextNode",
    "richText[0]",
  ],
  [
    "transformField",
    "richText[0].child",
  ],
  [
    "transformField",
    "richText[0].string",
  ],
  [
    "transformField",
    "valueItem",
  ],
  [
    "transformFieldItem",
    "valueItem",
  ],
  [
    "transformField",
    "valueItem.child",
  ],
  [
    "transformFieldItem",
    "valueItem.child",
  ],
  [
    "transformField",
    "valueItem.child.child",
  ],
  [
    "transformField",
    "valueItem.child.string",
  ],
  [
    "transformFieldItem",
    "valueItem.child.string",
  ],
  [
    "transformField",
    "valueItem.string",
  ],
  [
    "transformFieldItem",
    "valueItem.string",
  ],
  [
    "transformField",
    "valueItemList",
  ],
  [
    "transformFieldItem",
    "valueItemList[0]",
  ],
  [
    "transformField",
    "valueItemList[0].child",
  ],
  [
    "transformField",
    "valueItemList[0].string",
  ],
  [
    "transformFieldItem",
    "valueItemList[0].string",
  ],
  [
    "transformFieldItem",
    "valueItemList[1]",
  ],
  [
    "transformField",
    "valueItemList[1].child",
  ],
  [
    "transformField",
    "valueItemList[1].string",
  ],
  [
    "transformFieldItem",
    "valueItemList[1].string",
  ],
]
`;
